### PR TITLE
Custom Log Format 

### PR DESCRIPTION
**Pull request by CharityKathure**
**Summary**

In addition to getting logs in XML and JSON formats, a user should have the flexibility to customize their windows container logs, dictating how the logs should look like and the specific fields that should appear in log entries.  

For Example: 
To display the timestamp, source, severity and message of an event in STDOUT, a user can specify the following format: 
`[%TimeStamp%] [%Source%] [%Severity%] %Message% `

**Examples:**

**1. Event Logs**
**Configuration:** `"customLogFormat": "[%TimeStamp%] [%Source%] [%Severity%] %Message%"` 
**Output:** _
```
[2023-02-17T02:35:42.000Z] [EventLog] [Information] Successfully logged OS information
```

**2. ETW Logs**
**Configuration:** ` "customLogFormat": "[%TimeStamp%] [%Source%] [%Severity%] [%ProviderId%] [%ProviderName%] [%EventId%] %EventData%"`
**Output:** _
```
[2023-03-10T07:27:59.000Z] [ETW] [Information] [{DAA6A96B-F3E7-4D4D-A0D6-31A350E6A445}] [Microsoft-Windows-WLAN-Driver] [1] FrameUniqueID: 462621 QueueLength: 0 QueueState: false Status: 0 CustomData1: 0 CustomData2: 0 CustomData3: 0
```

**3. File Logs**
**Configuration:** `"[%TimeStamp%] [%Source%] [%FileName%] %Message%"`
**Output:** 
```
[2023-03-10T07:31:08.000Z] [File] [Trial.log] Testing file log entry on STDOUT
```

**Sample Config used in testing:**
```json
{
  "LogConfig": {
    "logFormat": "custom",
    "sources": [
      {
        "type": "File",
        "directory": "c:\\inetpub\\logs",
        "filter": "*.log",
        "includeSubdirectories": true,
        "customLogFormat": "[%TimeStamp%] [%Source%] [%FileName%] %Message%"
      },
      {
        "type": "ETW",
        "eventFormatMultiLine": false,
        "providers": [
          {
            "providerName": "Microsoft-Windows-WLAN-Drive",
            "providerGuid": "DAA6A96B-F3E7-4D4D-A0D6-31A350E6A445",
            "level": "Information"
          }
        ],
        "customLogFormat": "[%TimeStamp%] [%Source%] [%Severity%] [%ProviderId%] [%ProviderName%] [%DecodingSource%] [%ExecutionProcessId%] [%ExecutionThreadId%] [%EventId%] %eventdata%"
      }
    ]
  }
}
```

### CODE REVIEW

File: LogMonitor/src/LogMonitor/EventMonitor.cpp [hunk 1]
```diff
@@ -554,30 +560,36 @@ EventMonitor::PrintEvent(
...
+                pLogEntry->eventLevel = c_LevelToString[static_cast<UINT8>(level)];
+                pLogEntry->eventMessage = (LPWSTR)(&m_eventMessageBuffer[0]);
+
+                std::wstring formattedEvent;
+                if (Utility::CompareWStrings(m_logFormat, L"Line")) {
+                    formattedEvent = FormatEventLineLog(m_lineLogFormat, pLogEntry);
                 }
+                else {
+                    std::wstring logFmt = L"<Log><Source>%s</Source><LogEntry><Time>%s</Time><Channel>%s</Channel><Level>%s</Level><EventId>%u</EventId><Message>%s</Message></LogEntry></Log>";
+                    if (Utility::CompareWStrings(m_logFormat, L"JSON"))
```
[the remark this answers is not in the document]
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/EventMonitor.cpp:575]: They are currently on different level because JSON and XML formatting utilize the Utility::FormatString() function, which is not required in Line formatting. 

Putting all the `if`s on the same level will lead to repetition of this bit for bot XML and JSON:
```
formattedEvent = Utility::FormatString(
                        logFmt.c_str(),
                        pLogEntry->source.c_str(),
                        pLogEntry->eventTime.c_str(),
                        pLogEntry->eventChannel.c_str(),
                        pLogEntry->eventLevel.c_str(),
                        pLogEntry->eventId,
                        pLogEntry->eventMessage.c_str()
                    );
```

**Review by bobsira — COMMENTED** [review]
Leaving these comment for ETW changes atm!

File: LogMonitor/src/LogMonitor/EtwMonitor.cpp [hunk 2]
```diff
@@ -30,9 +30,11 @@ static const std::wstring g_sessionName = L"Log Monitor ETW Session";
 
 EtwMonitor::EtwMonitor(
     _In_ const std::vector<ETWProvider>& Providers,
-    _In_ std::wstring LogFormat
+    _In_ std::wstring LogFormat,
+    _In_ std::wstring LineLogFormat
     ) :
-    m_logFormat(LogFormat)
+    m_logFormat(LogFormat),
+    m_lineLogFormat(LineLogFormat)
```
[the remark this answers is not in the document]
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/EtwMonitor.cpp:37]: Yes. There is a default configuration for each source type. _(in LoggerSettings.h)_

File: LogMonitor/src/LogMonitor/EventMonitor.cpp [hunk 3]
```diff
@@ -722,3 +734,39 @@ EventMonitor::EnableEventLogChannel(
         EvtClose(channelConfig);
     }
 }
+
+std::wstring EventMonitor::EventFieldsMapping(_In_ std::wstring eventFields, _Inout_ EventLogEntry* pLogEntry)
```
[the remark this answers is not in the document]
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/EventMonitor.cpp:737]: The mapping function serves to map the fields as obtained from the ''linelogformat'' field in the config file during decoding with the struct values

File: LogMonitor/src/LogMonitor/LogFileMonitor.cpp [hunk 4]
```diff
@@ -2045,3 +2069,38 @@ LogFileMonitor::GetFileId(
 
     return status;
 }
+
+std::wstring LogFileMonitor::FileFieldsMapping(_In_ std::wstring fileFields, _Inout_ FileLogEntry* pLogEntry)
```
[the remark this answers is not in the document]
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/LogFileMonitor.cpp:2072]: The mapping function serves to map the fields as obtained from the ''linelogformat'' field in the config file during decoding with the struct values

File: LogMonitor/src/LogMonitor/EtwMonitor.h [hunk 5]
```diff
@@ -34,7 +35,8 @@ class EtwMonitor final
 
     EtwMonitor(
         _In_ const std::vector<ETWProvider>& Providers,
-        _In_ std::wstring LogFormat
+        _In_ std::wstring LogFormat,
+        _In_ std::wstring LineLogFormat
```
[the remark this answers is not in the document]
Comment (profnandaa) [reply LogMonitor/src/LogMonitor/EtwMonitor.h:39]: Ok, got it from our discussion on the call.

File: LogMonitor/src/LogMonitor/EventMonitor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (profnandaa) [reply LogMonitor/src/LogMonitor/EventMonitor.cpp:574]: Understood from your clarification on the call. I think I'm okay with avoiding the repetition.

File: LogMonitor/src/LogMonitor/LogFileMonitor.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/LogFileMonitor.cpp]: Function Moved to Utility.cpp to avoid code repetition.

File: LogMonitor/src/LogMonitor/EventMonitor.cpp [hunk 6]
```diff
@@ -488,7 +496,7 @@ EventMonitor::PrintEvent(
                 status = GetLastError();
 
                 logWriter.TraceError(
-                    Utility::FormatString(L"Failed to render event. Error: %lu", status).c_str()
+                    util.FormatString(L"Failed to render event. Error: %lu", status).c_str()
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/EventMonitor.cpp:499]: why the change form `Utility` from the _static_ approach we've been using?

File: LogMonitor/src/LogMonitor/Utility.cpp [hunk 7]
```diff
@@ -324,4 +325,39 @@ bool Utility::CompareWStrings(wstring stringA, wstring stringB)
...
+    size_t i = 0, j = 1;
+    while (i < customLogFormat.size()) {
+
+        auto sub = customLogFormat.substr(i, j - i);
+        auto sub_length = sub.size();
+
+        if (sub[0] != '%' && sub[sub_length - 1] != '%') {
+            j++, i++;
+        } else if (sub[0] == '%' && sub[sub_length - 1] == '%' && sub_length != 1) {
+
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/Utility.cpp:341]: nit: can remove the empty line.

File: LogMonitor/src/LogMonitor/Utility.cpp [hunk 8]
```diff
@@ -324,4 +325,39 @@ bool Utility::CompareWStrings(wstring stringA, wstring stringB)
                 return towupper(l1) == towupper(r1);
             }
         );
+}
+
+std::wstring Utility::FormatEventLineLog(_In_ std::wstring customLogFormat, _In_ void* pLogEntry, _In_ std::wstring sourceType)
+{
+    size_t i = 0, j = 1;
+    while (i < customLogFormat.size()) {
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/Utility.cpp:333]: I wanted to suggest that we could use `<regex>` instead of the custom approach but looks like this approach will be shorter and perhaps more efficient 👍

File: LogMonitor/LogMonitorTests/EtwMonitorTests.cpp [hunk 9]
```diff
@@ -203,7 +203,7 @@ namespace LogMonitorTests
             fflush(stdout);
 
 
-            std::function<void(void)> f1 = [&etwProviders] { EtwMonitor etwMonitor(etwProviders, L"json"); };
+            std::function<void(void)> f1 = [&etwProviders] { EtwMonitor etwMonitor(etwProviders, L"json", L""); };
```
Comment (profnandaa) [inline LogMonitor/LogMonitorTests/EtwMonitorTests.cpp:206]: nit: could the second param just be optional with `""` default, instead of passing it here?

**Review by profnandaa — COMMENTED** [review]
LGTM for the most part, just a few comments. And I also see builds are failing, I don't know why.

File: LogMonitor/src/LogMonitor/EventMonitor.cpp (hunk 6 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/EventMonitor.cpp]: This change was to enable reference a non-static method (FormatEventLineLog) but I have made method static and reverted the changes done one the previous approach.

File: LogMonitor/src/LogMonitor/Utility.cpp (hunk 7 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/Utility.cpp]: Empty line removed.

File: LogMonitor/src/LogMonitor/Utility.cpp (hunk 8 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/Utility.cpp:333]: Thank you.

File: LogMonitor/LogMonitorTests/EtwMonitorTests.cpp (hunk 9 above)
Comment (CharityKathure) [reply LogMonitor/LogMonitorTests/EtwMonitorTests.cpp:206]: @profnandaa It was a trade-off between creating an overloaded function that takes 2 parameters or assigning default value to CustomLogFormat.

File: LogMonitor/src/LogMonitor/EtwMonitor.cpp [hunk 10]
```diff
@@ -30,9 +30,11 @@ static const std::wstring g_sessionName = L"Log Monitor ETW Session";
 
 EtwMonitor::EtwMonitor(
     _In_ const std::vector<ETWProvider>& Providers,
-    _In_ std::wstring LogFormat
+    _In_ std::wstring LogFormat,
+    _In_ std::wstring CustomLogFormat
```
Comment (profnandaa) [inline LogMonitor/src/LogMonitor/EtwMonitor.cpp:34]: I meant having `CustomLogFormat = L""` instead.

File: LogMonitor/LogMonitorTests/EtwMonitorTests.cpp (hunk 9 above)
Comment (profnandaa) [reply LogMonitor/LogMonitorTests/EtwMonitorTests.cpp:206]: Sorry, I meant modifying Line 35 of EtwMonitor constructor.

**Review by profnandaa — COMMENTED** [review]
Just one minor and I'm good.

File: LogMonitor/src/LogMonitor/EtwMonitor.cpp (hunk 10 above)
Comment (CharityKathure) [reply LogMonitor/src/LogMonitor/EtwMonitor.cpp:34]: @profnandaa even with making this change at the constructor level,  three variables will still need to be passed in the test cases.

**Review by profnandaa — APPROVED** [review]

**Review by bobsira — APPROVED** [review]
Tested the change on host and on a container and everything looks good on my end!
